<h1> Encrypted Inference-Linear Regression</h1>

Author:
<ul>
    <li>Hrishikesh Kamath - <a href="https://twitter.com/kamathhrishi">Twitter</a> - <a href="https://github.com/kamathhrishi">Github</a>
</ul>
Encrypted Inference is the process of performing inference with machine learning models such that model owner cannot observe the true input data nor can the data owner see the true model weights. The weights and data are encrypted by splitting them into shares and performiming computations according to a protocol. The general class of methods know as <b>Secure Multi Party Computation (SMPC)</b>. 

Below figure depicts MPC for ML models for 2 parties. 

<img height="600px" width="600px" src="Images/smpc_illustration.png"></img>

In this example we use Virtual Machine to demonstrate performing inference using SMPC. That is the workers are present on the same PC. If you want to understand how to perform remotely check out duet tutorials. 
In this example, we train a Linear regression model in plaintext on Boston Housing Dataset. Then we use the model for performing encrypted inference on test data. This tutorial uses protocol Falcon for 3 parties and SPDZ for 3 and 5 parties. 

In SyMPC the computation between parties occurs using a orchestrator which describes how computations should take place.

In [1]:
#External libraries
import pandas as pd
import numpy as np
import time

In [2]:
#Import torch
import torch
import torch.nn as nn
import torch.utils.data as data_utils

In [3]:
#Set a manual seed to maintain consistency
torch.manual_seed(0)

<h2>Data Loading and Processing</h2>

In [4]:
#Download Boston housing dataset
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data

--2021-07-22 20:30:51--  https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 49082 (48K) [application/x-httpd-php]
Saving to: ‘housing.data.14’

housing.data.14     100%[===================>]  47.93K  93.7KB/s    in 0.5s    

2021-07-22 20:30:52 (93.7 KB/s) - ‘housing.data.14’ saved [49082/49082]



In [5]:
#Import dataset and add headers
dataset=pd.read_csv("housing.data",delim_whitespace=True,
                    names=["crim","zn","indus",
                           "chas","nox","rm",
                           "age","dis","rad",
                           "tax","ptratio","black",
                           "lstat","medv"])

In [6]:
#Visualize and look at columns and rows of dataset
dataset.head()

crim    zn  indus  chas    nox     rm   age     dis  rad    tax  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   ptratio   black  lstat  medv  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [7]:
#Split data into features and target variables
features = dataset.drop("medv",axis=1)
targets = dataset["medv"]

In [8]:
#Normalize features
features = features.apply(
    lambda x: (x - x.mean()) / x.std()
)

In [9]:
features

crim        zn     indus      chas       nox        rm       age  \
0   -0.419367  0.284548 -1.286636 -0.272329 -0.144075  0.413263 -0.119895   
1   -0.416927 -0.487240 -0.592794 -0.272329 -0.739530  0.194082  0.366803   
2   -0.416929 -0.487240 -0.592794 -0.272329 -0.739530  1.281446 -0.265549   
3   -0.416338 -0.487240 -1.305586 -0.272329 -0.834458  1.015298 -0.809088   
4   -0.412074 -0.487240 -1.305586 -0.272329 -0.834458  1.227362 -0.510674   
..        ...       ...       ...       ...       ...       ...       ...   
501 -0.412820 -0.487240  0.115624 -0.272329  0.157968  0.438881  0.018654   
502 -0.414839 -0.487240  0.115624 -0.272329  0.157968 -0.234316  0.288648   
503 -0.413038 -0.487240  0.115624 -0.272329  0.157968  0.983986  0.796661   
504 -0.407361 -0.487240  0.115624 -0.272329  0.157968  0.724955  0.736268   
505 -0.414590 -0.487240  0.115624 -0.272329  0.157968 -0.362408  0.434302   

          dis       rad       tax   ptratio     black     lstat  
0    0.140075 -0.981871 -0.665949 -1.457558  0.440616 -1.074499  
1    0.556609 -0.867024 -0.986353 -0.302794  0.440616 -0.491953  
2    0.556609 -0.867024 -0.986353 -0.302794  0.396035 -1.207532  
3    1.076671 -0.752178 -1.105022  0.112920  0.415751 -1.360171  
4    1.076671 -0.752178 -1.105022  0.112920  0.440616 -1.025487  
..        ...       ...       ...       ...       ...       ...  
501 -0.625178 -0.981871 -0.802418  1.175303  0.386834 -0.417734  
502 -0.715931 -0.981871 -0.802418  1.175303  0.440616 -0.500355  
503 -0.772919 -0.981871 -0.802418  1.175303  0.440616 -0.982076  
504 -0.667776 -0.981871 -0.802418  1.175303  0.402826 -0.864446  
505 -0.612640 -0.981871 -0.802418  1.175303  0.440616 -0.668397  

[506 rows x 13 columns]

In [10]:
targets

0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
       ... 
501    22.4
502    20.6
503    23.9
504    22.0
505    11.9
Name: medv, Length: 506, dtype: float64

In [11]:
#Convert features and targets into torch tensors
features = torch.tensor(features.values.astype(np.float32)) 
targets = torch.tensor(targets.values.astype(np.float32))

In [12]:
# Arguments for training
batch_size = 16
epochs = 300
train_test_split = 0.8
lr = 0.001

In [13]:
#Split dataset into train and test
train_indices=int(len(features)*train_test_split)

train_x = features[:train_indices]
train_y = targets[:train_indices]

test_x = features[train_indices+1:]
test_y = targets[train_indices+1:]

In [14]:
#Divide dataset into batches
def get_batches(X, y):
    batches = []
    for index in range(0,len(train_x)+1,batch_size):
        batches.append((X[index:index+batch_size],y[index:index+batch_size]))
    
    return batches

In [15]:
train_batches=get_batches(train_x,train_y)

<h1>Plaintext Training</h1>

In [16]:
#Import syft
import syft as sy
sy.logger.remove()

In [17]:
#Define Linear regression model
class LinearSyNet(sy.Module):
    def __init__(self, torch_ref):
        super(LinearSyNet, self).__init__(torch_ref=torch_ref)
        self.fc1 = self.torch_ref.nn.Linear(13,1)

    def forward(self, x):
        x = self.fc1(x)
        return x

In [18]:
#Define model, loss function and optimizer
model = LinearSyNet(torch)
criterion = torch.nn.MSELoss(reduction='mean') 
optimizer = torch.optim.SGD(model.parameters(), lr=lr)

In [19]:
#Training Loop
for epoch in range(epochs):
  running_loss = 0.0
  for index in range(0,len(train_batches)):
    # Clear gradient buffers because we don't want any gradient from previous epoch to carry forward, dont want to cummulate gradients
    optimizer.zero_grad()

    # get output from the model, given the inputs
    outputs = model(train_batches[index][0]).reshape([-1])

    # get loss for the predicted output
    loss = criterion(outputs,train_batches[index][1])
    running_loss += loss
    # get gradients w.r.t to parameters
    loss.backward()

    # update parameters
    optimizer.step()
    
  test_accuracy = criterion(model(test_x).reshape([-1]),test_y)
  if((epoch%50)==0):
     print(f"Epoch {epoch}/{epochs}  Running Loss : {running_loss.item()/batch_size} and test loss: {test_accuracy.item()}")

Epoch 0/300  Running Loss : 978.0777587890625 and test loss: 367.94952392578125
Epoch 50/300  Running Loss : 52.76874542236328 and test loss: 108.14092254638672
Epoch 100/300  Running Loss : 43.0278434753418 and test loss: 47.971736907958984
Epoch 150/300  Running Loss : 41.08037185668945 and test loss: 30.054576873779297
Epoch 200/300  Running Loss : 40.123294830322266 and test loss: 23.0802059173584
Epoch 250/300  Running Loss : 39.57941436767578 and test loss: 20.64508628845215


<h1>Plaintext Inference</h1>

In [20]:
#Perform inference in plaintext
start_time=time.time()
plaintext_predictions = model(test_x).reshape([-1])
end_time=time.time()

In [21]:
#Calculate inference time and MSELoss
print("MSE Loss: ",criterion(plaintext_predictions,test_y).item())
print("Inference time: ",str(end_time-start_time),"s")

MSE Loss:  20.227853775024414
Inference time:  0.0005340576171875 s


<h1>Encrypted Inference</h1>

In [22]:
#SyMPC imports required for encrypted inference
import sympc
from sympc.session import Session
from sympc.session import SessionManager
from sympc.tensor import MPCTensor
from sympc.protocol import Falcon,FSS

In [23]:
def get_clients(n_parties):
  #Generate required number of syft clients and return them.

  parties=[]
  for index in range(n_parties): 
      parties.append(sy.VirtualMachine(name = "worker"+str(index)).get_root_client())

  return parties

In [24]:
def inference(n_clients,protocol=None):
    
  # Get VM clients 
  parties=get_clients(n_clients)

  # Setup the session for the computation
  if(protocol):
     session = Session(parties = parties,protocol = protocol)
  else:
     session = Session(parties = parties)
        
  SessionManager.setup_mpc(session)

  #Encrypt model 
  mpc_model = model.share(session)

  #Encrypt test data
  test_data=MPCTensor(secret=test_x, session = session)

  #Perform inference and measure time taken
  start_time = time.time()
  enc_results = mpc_model(test_data)
  end_time = time.time()

  print(f"Time for inference: {end_time-start_time}s")

  #Get plaintext predictions
  predictions = enc_results.reconstruct()
  
  #Calculate Loss
  print("MSE Loss: ",criterion(predictions.reshape([-1]),test_y).item())
    
  return predictions

In [25]:
predictions=inference(3,Falcon("semi-honest"))

Time for inference: 0.03487205505371094s
MSE Loss:  20.227659225463867


We can see that the prediction values and mean squared error values are almost the same as final model. Small differences are due to precision loss.

In [26]:
for index in range(0,10):
    print(f"Index {index}")
    print(f"Encrypted Prediction Output {predictions[index].item()}")
    print(f"Plaintext Prediction Output {plaintext_predictions[index].item()}")
    print(f"Expected Prediction: {test_y[index]}")
    print("\n")

Index 0
Encrypted Prediction Output 2.6369781494140625
Plaintext Prediction Output 2.6367740631103516
Expected Prediction: 5.0


Index 1
Encrypted Prediction Output 4.957275390625
Plaintext Prediction Output 4.957102298736572
Expected Prediction: 11.899999618530273


Index 2
Encrypted Prediction Output 19.214553833007812
Plaintext Prediction Output 19.214506149291992
Expected Prediction: 27.899999618530273


Index 3
Encrypted Prediction Output 12.028594970703125
Plaintext Prediction Output 12.028535842895508
Expected Prediction: 17.200000762939453


Index 4
Encrypted Prediction Output 19.063995361328125
Plaintext Prediction Output 19.063936233520508
Expected Prediction: 27.5


Index 5
Encrypted Prediction Output 11.89825439453125
Plaintext Prediction Output 11.898107528686523
Expected Prediction: 15.0


Index 6
Encrypted Prediction Output 16.335174560546875
Plaintext Prediction Output 16.335037231445312
Expected Prediction: 17.200000762939453


Index 7
Encrypted Prediction Output -1.25

<h1> Conclusion </h1>

Falcon can also provide a malicious security guarantee for an honest majority at the cost of higher inference time. Malicious security ensures that all the parties compute according to the protocol and do not deviate from protocol or tamper with shares. 

In [27]:
predictions=inference(3,Falcon("malicious"))

Time for inference: 0.5890171527862549s
MSE Loss:  20.227663040161133


When we do not pass any protocol to session, SyMPC uses SPDZ and FSS protocol with semi-honest security type. 

SPDZ is used for multiplication and related operations (convolution,matmul,etc).
Functional Secret Sharing (FSS) for other operations such as comparison, equality, maxpool, etc. 

FSS works only for 2 parties while SPDZ could extend to N parties. 

Linear regression uses only matmul which utilizes SPDZ protocol allowing us to run linear regression with several parties in this tutorial. 

In [28]:
predictions=inference(3)

Time for inference: 0.5398709774017334s
MSE Loss:  20.227680206298828


In [29]:
predictions=inference(5)

Time for inference: 0.8621780872344971s
MSE Loss:  20.227649688720703


<center><h3> Comparison </h3></center>

| Protocol | Security Type| Parties | Inference Time (s) |
| --- | --- | --- | --- |
| Plaintext | |  | 0.000534|
| Falcon | Semi-honest | 3 | 0.0348 |
| Falcon | Malicious | 3 | 0.5890 |
| SPDZ| Semi-honest | 3 | 0.5274 |
| SPDZ | Semi-honest | 5 | 0.8622|

<b>Note:</b> The above table is only for comparison. The inference time varies for different PC specs and CPU load.



Falcon provides faster inference for a 3 parties setting. While, SPDZ allows inference for N number of parties. Both allow inference with almost same accuracy as plaintext. 

Although Falcon is much more faster, it isn't scalable because it is applicable for only 3 parties. Further, it is less secure. Since it uses 2-out-of-3 sharing where each party recieves two shares allowing 2 parties to reconstruct a secret without other party knowing. Falcon also assumes that majority of parties are honest (2 in this case). 

While, SPDZ and FSS distributes a single share to every party requiring shares from all parties for reconstruction ensuring no parties could collude. Currently SyMPC provides support for SPDZ and FSS with semi-honest security guarantee. This allows parties to tamper with shares leading to incorrect results. 

<h3>What's next?</h3>

SyMPC is still under development! We will add here more features as soon they are stable enough, stay tuned! 🕺

If you enjoyed this tutorial, show your support by Starring SyMPC! 🙏